src/bin/pg_autoctl/service_postgres_ctl.c

```diff
@@ -0,0 +1,511 @@
+{
+	Program program;
+
+	char *args[12];
+	int argsIndex = 0;
+
+	char command[BUFSIZE];
+
+	/*
+	 * use --pgdata option rather than the config.
+	 *
+	 * On macOS when using /tmp, the file path is then redirected to being
+	 * /private/tmp when using realpath(2) as we do in normalize_filename(). So
+	 * for that case to be supported, we explicitely re-use whatever PGDATA or
+	 * --pgdata was parsed from the main command line to start our sub-process.
+	 *
+	 * The pg_autoctl postgres controller is used both in the monitor context
+	 * and in the keeper context; which means it gets started from one of the
+	 * following top-level commands:
+	 *
+	 *  - pg_autoctl create monitor
+	 *  - pg_autoctl create postgres
+	 *  - pg_autoctl run
+	 *
+	 * The monitor specific commands set monitorOptions, the generic and keeper
+	 * specific commands set keeperOptions.
+	 */
+	char *pgdata =
+		IS_EMPTY_STRING_BUFFER(monitorOptions.pgSetup.pgdata)
+		? keeperOptions.pgSetup.pgdata
+		: monitorOptions.pgSetup.pgdata;
+
+	setenv(PG_AUTOCTL_DEBUG, "1", 1);
+
+	args[argsIndex++] = (char *) pg_autoctl_program;
+	args[argsIndex++] = "do";
+	args[argsIndex++] = "service";
+	args[argsIndex++] = "postgres";
+	args[argsIndex++] = "--pgdata";
+	args[argsIndex++] = pgdata;
+	args[argsIndex++] = logLevelToString(log_get_level());
+	args[argsIndex] = NULL;
+
+	/* we do not want to call setsid() when running this program. */
+	program = initialize_program(args, false);
+
+	program.capture = false;    /* redirect output, don't capture */
+	program.stdOutFd = STDOUT_FILENO;
+	program.stdErrFd = STDERR_FILENO;
+
+	/* log the exact command line we're using */
+	(void) snprintf_program_command_line(&program, command, BUFSIZE);
+
+	log_info("%s", command);
+
+	(void) execute_program(&program);
+}
+
+
+/*
+ * service_postgres_ctl_loop loops over the current CTL state file and ensure
+ * that Postgres is running when that's expected, or that Postgres is not
+ * running when in a state where we should maintain Postgres down to avoid
+ * split-brain situations.
+ */
+void
+service_postgres_ctl_loop(LocalPostgresServer *postgres)
+{
+	PostgresSetup *pgSetup = &(postgres->postgresSetup);
+	LocalExpectedPostgresStatus *localStatus = &(postgres->expectedPgStatus);
+	KeeperStatePostgres *pgStatus = &(localStatus->state);
+
+	Service postgresService = {
+		"postgres",
+		-1,
+		&service_postgres_start,
+		&service_postgres_stop,
+		&service_postgres_reload,
+		(void *) pgSetup
+	};
+
+	bool pgStatusPathIsReady = false;
+
+	/* make sure to initialize the expected Postgres status to unknown */
+	pgStatus->pgExpectedStatus = PG_EXPECTED_STATUS_UNKNOWN;
+
+	for (;;)
+	{
+		pid_t pid;
+		int status;
+
+		/* we might have to reload, pass the signal down */
+		if (asked_to_reload)
+		{
+			(void) service_postgres_reload((void *) &postgresService);
+
+			asked_to_reload = 0;
+		}
+
+		/* that's expected the shutdown sequence from the supervisor */
+		if (asked_to_stop || asked_to_stop_fast)
```

Comment:
Well Postgres stop modes aren't as useful as they look, I think it's okay to always use fast. At least we should never use immediate, of course, because we don't want to lose data. And the smart mode is all but.

Reply:
Okay that makes sense. Especially since with HA, you should be able to handle retries of queries.